hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java

```diff
@@ -770,6 +776,214 @@ public void testLookupKeyWithLocation() throws IOException {
+    String prefixKeyInDB = "key-d";
+    String prefixKeyInCache = "key-c";
+    for (int i = 1; i <= 10; i++) {
+      if (i % 2 == 0) {  // Add to DB
+        TestOMRequestUtils.addKeyToTable(false,
+            VOLUME_NAME, BUCKET_NAME,
+            keyNameDir1Subdir1 + OZONE_URI_DELIMITER + prefixKeyInDB + i,
+            1000L, HddsProtos.ReplicationType.RATIS,
+            HddsProtos.ReplicationFactor.ONE, metadataManager);
+      } else {  // Add to TableCache
+        TestOMRequestUtils.addKeyToTableCache(
+            VOLUME_NAME, BUCKET_NAME,
+            keyNameDir1Subdir1 + OZONE_URI_DELIMITER + prefixKeyInCache + i,
+            HddsProtos.ReplicationType.RATIS, HddsProtos.ReplicationFactor.ONE,
+            metadataManager);
+      }
+    }
+
+    // Test non-recursive, should return the dir under root
+    fileStatuses =
+        keyManager.listStatus(rootDirArgs, false, "", 1000);
+    Assert.assertEquals(2, fileStatuses.size());
+
+    // Test recursive, should return the dir and the keys in it
+    fileStatuses =
+        keyManager.listStatus(rootDirArgs, true, "", 1000);
+    Assert.assertEquals(10 + 3, fileStatuses.size());
+
+    // Clean up
+    for (int i = 1; i <= 10; i += 2) {
+      // Mark TableCache entries as deleted
+      // Note that DB entry clean up is handled by cleanupTest()
+      String key = metadataManager.getOzoneKey(
+          VOLUME_NAME, BUCKET_NAME,
+          keyNameDir1Subdir1 + OZONE_URI_DELIMITER + prefixKeyInCache + i);
+      metadataManager.getKeyTable().addCacheEntry(new CacheKey<>(key),
+          new CacheValue<>(Optional.absent(), 2L));
+    }
+  }
+
+  @Test
+  public void testListStatusWithDeletedEntriesInCache() throws Exception {
+    String prefixKey = "key-";
+    TreeSet<String> existKeySet = new TreeSet<>();
+    TreeSet<String> deletedKeySet = new TreeSet<>();
+
+    for (int i = 1; i <= 100; i++) {
+      if (i % 2 == 0) {
+        TestOMRequestUtils.addKeyToTable(false,
+            VOLUME_NAME, BUCKET_NAME, prefixKey + i,
+            1000L, HddsProtos.ReplicationType.RATIS,
+            HddsProtos.ReplicationFactor.ONE, metadataManager);
+        existKeySet.add(prefixKey + i);
+      } else {
+        TestOMRequestUtils.addKeyToTableCache(
+            VOLUME_NAME, BUCKET_NAME, prefixKey + i,
+            HddsProtos.ReplicationType.RATIS, HddsProtos.ReplicationFactor.ONE,
+            metadataManager);
+
+        String key = metadataManager.getOzoneKey(
+            VOLUME_NAME, BUCKET_NAME, prefixKey + i);
+        // Mark as deleted in cache.
+        metadataManager.getKeyTable().addCacheEntry(new CacheKey<>(key),
+            new CacheValue<>(Optional.absent(), 2L));
+        deletedKeySet.add(key);
+      }
+    }
+
+    OmKeyArgs rootDirArgs = createKeyArgs("");
+    List<OzoneFileStatus> fileStatuses =
+        keyManager.listStatus(rootDirArgs, true, "", 1000);
+    // Should only get entries that are not marked as deleted.
```

Comment:
Can we add some tests for the following cases?
1. keyName="someValue", startKey="", count="n" and after that call list status again with next 10 files to get.
2. Can we also have some test for multi-level directory and test above scenario?

Reply:
Sure. Added pagination check to existing test (with cache).

Also `testListStatus()` has pagination test (without cache) with two-level deep dir.